### PR TITLE
feat: UI selector for project/dataset

### PR DESCRIPTION
**Pull request by SanjeevLakhwani**
Follows from #161
- improved over the existing redux over the metadata, reducing more stateful property
- Found a bug with `/callback` in index.js, it was never being triggered as a specific route as it was being accessed before by :lang param
- provided routing for project/dataset navigation by the format of /p/projecId/d/datasetId, rendered to the level needed
- modified the ui to select project/dataset 
- improvided on app structure by having separation of concerns with components and using Outlets and Loaders

### CODE REVIEW

File: src/js/components/ChooseProjectModal.tsx [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                <Typography.Text>{description}</Typography.Text>
+                <Typography.Title level={5}>Datasets</Typography.Title>
+                <List
+                  dataSource={datasets}
+                  bordered
+                  renderItem={(item) => (
+                    <Link to={`${baseURL}/p/${identifier}/d/${item.identifier}`}>
+                      <List.Item className="select-dataset-hover" key={item.identifier}>
+                        <List.Item.Meta
+                          avatar={<Avatar src="/public/assets/database.png" />}
```
Comment (davidlougheed) [inline src/js/components/ChooseProjectModal.tsx:57]: ideally this should be an existing icon / component, and at the very least an SVG

File: src/js/components/ChooseProjectModal.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
+import React, { useState } from 'react';
+import { Tabs, List, Avatar, Modal, Button, Space, Typography } from 'antd';
+import { useAppSelector } from '@/hooks';
+import { Link, useLocation } from 'react-router-dom';
+
+const ChooseProjectModal: React.FC<ChooseProjectModalProps> = ({ isModalOpen, setIsModalOpen }) => {
```
Comment (davidlougheed) [inline src/js/components/ChooseProjectModal.tsx:6]: React.FC is not ideal as a type: see https://github.com/facebook/create-react-app/pull/8177

File: src/js/components/ChooseProjectModal.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            <Button>Select Default View</Button>
+          </Link>
+        </Space>
+      }
+    >
+      <Tabs
+        tabPosition="left"
+        activeKey={selectedProject}
+        onChange={(key) => setSelectedProject(key)}
+        items={projects.map(({ identifier, title, datasets, description }) => {
```
Comment (davidlougheed) [inline src/js/components/ChooseProjectModal.tsx:36]: I'm not a big fan of the way these are rendered - the visual hierarchy is unclear, as is the "cancel" button's purpose and the "select default view" button.

**Review by davidlougheed — CHANGES_REQUESTED** [review]
please fix merge conflicts.

File: src/js/components/ChooseProjectModal.tsx (hunk 1 above)
Comment (SanjeevLakhwani) [reply src/js/components/ChooseProjectModal.tsx:57]: I am not sure why it needs to be an svg. Given the rendering size of the icon, a png works fine.

File: src/js/components/ChooseProjectModal.tsx (hunk 3 above)
Comment (SanjeevLakhwani) [reply src/js/components/ChooseProjectModal.tsx:36]: How about this: 
image

Comment (gsfk) [reply src/js/components/ChooseProjectModal.tsx:27]: As a sometimes confused user, I'm still a bit puzzled at the user actions needed to choose between viewing a whole project and just viewing a particular dataset. Intuitively you'd think they would be more similar.

File: src/styles.css [hunk 4]
```diff
@@ -22,3 +22,12 @@ body {
 .ant-layout-sider-trigger {
   border-right: 1px solid #f0f0f0;
 }
+
+.select-project-title:hover {
+  color: white !important;
```
Comment (davidlougheed) [inline src/styles.css:27]: add a transition for the color change

File: src/styles.css [hunk 5]
```diff
@@ -22,3 +22,12 @@ body {
...
+  cursor: pointer;
+}
+
+.select-dataset-hover:hover {
+  background-color: aliceblue !important;
```
Comment (davidlougheed) [inline src/styles.css:32]: add a transition for the color change (~0.15s)

File: src/styles.css (hunk 4 above)
Comment (davidlougheed) [reply src/styles.css:27]: (maybe ~0.15s)

**Review by davidlougheed — CHANGES_REQUESTED** [review]
image

image


still many rendering glitches here:

* unneeded margin above project title
* project title and datasets subheading should not be the same heading level
* button margin for "clear selection"
* hover background for dataset goes outside of border radius bounds
* text in header isn't aligned with icon vertically
* I think the node heading should be a bit bigger, and then so should this selector heading in the header

**Comment by davidlougheed** [comment]
image
i also get this when I load – no routes match `/`

**Comment by davidlougheed** [comment]
image

File: src/js/components/ChooseProjectModal.tsx [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+import { useAppSelector, useTranslationCustom, useTranslationDefault } from '@/hooks';
+import { Link, useLocation } from 'react-router-dom';
+import { FaDatabase } from 'react-icons/fa';
+
+const ChooseProjectModal = ({ isModalOpen, setIsModalOpen }: ChooseProjectModalProps) => {
+  const td = useTranslationDefault();
+  const t = useTranslationCustom();
+  const location = useLocation();
+  const { projects, selectedScope } = useAppSelector((state) => state.metadata);
+  const [selectedProject, setSelectedProject] = useState(selectedScope.project ?? projects[0].identifier);
```
Comment (davidlougheed) [inline src/js/components/ChooseProjectModal.tsx:12]: breaks with no projects (projects[0] is undefined)

File: src/js/components/SiteHeader.tsx [hunk 7]
```diff
@@ -40,16 +58,28 @@ const SiteHeader = () => {
...
+            level={5}
+            style={{ fontSize: '16px', margin: 0, lineHeight: '64px', color: 'lightgray' }}
+            onClick={() => setIsModalOpen(true)}
+          >
+            <ProfileOutlined style={{ marginRight: '5px', fontSize: '16px' }} />
+
+            {selectedScope.project && scopeProps.projectTitle}
+            {scopeProps.datasetTitle ? ` / ${scopeProps.datasetTitle}` : ''}
+          </Typography.Title>
+          )
```
Comment (davidlougheed) [inline src/js/components/SiteHeader.tsx:81]: stray ')'

**Review by davidlougheed — CHANGES_REQUESTED** [review]

File: src/js/types/metadata.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
+interface DiscoveryOverview {}
+interface DiscoverySearch {}
+interface DiscoveryFields {}
+interface DiscoveryRules {}
```
Comment (davidlougheed) [inline src/js/types/metadata.ts:4]: what's the plan for these types? I though there were already types for this stuff?

File: src/js/types/metadata.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  description: string;
+  discovery: Discovery;
+  datasets: Dataset[];
+}
+
+export interface Dataset {
+  identifier: string;
+  title: string;
+  description: string;
+  discovery: Discovery;
```
Comment (davidlougheed) [inline src/js/types/metadata.ts:24]: these can be null, no?

File: src/js/types/metadata.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  search: DiscoverySearch[];
+  fields: DiscoveryFields;
+  rules: DiscoveryRules;
+}
+
+export interface Project {
+  identifier: string;
+  title: string;
+  description: string;
+  discovery: Discovery;
```
Comment (davidlougheed) [inline src/js/types/metadata.ts:16]: these can be null, no?

**Review by davidlougheed — CHANGES_REQUESTED** [review]

**Review by davidlougheed — APPROVED** [review]
lgtm

**Pull request merged**